### PR TITLE
MM-27918 In-Product notices support

### PR DESCRIPTION
**Pull request by reflog**
#### Summary
This PR introduces support for In-Product Notices as described in https://docs.google.com/document/d/1LiEuv95fws__sjULbDkLvzj83U3yH_s2zk5FnMkd9tk/edit

It contains the following:

1. A new table storing 'viewed' state for each userid/noticeid pair
2. New job-server task that once an hour does the following:
 a) fetches new notices from an external server (waiting for DevOps to bring it up, in the meantime - uses public github repo. I'm leaving Do-Not-Merge label until the actual host is up and running)
 b) cleans the 'viewed' state for notices that are no longer existing
 c) calculates user and post count and stores them, since they are used for filtering notices
3. New API endpoints to 
 a) fetch notices appropriate for user
 b) allow user to mark notice as viewed
4. Two new settings AdminNoticesEnabled and UserNoticesEnabled , set to true by default

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-27918

### Related PRs
enterprise -https://github.com/mattermost/enterprise/pull/763
webapp - https://github.com/mattermost/mattermost-webapp/pull/6426

### CODE REVIEW

File: vendor/modules.txt
Comment (jespino) [inline vendor/modules.txt:405]: You need to add the changes to include in the NOTICE.txt file the licenses that applies to this new packages included. This one and the `Masterminds/semver/v3` one.

File: app/product_notices.go
Comment (ashishbhate) [inline app/product_notices.go:36]: `locale` is passed to this function but doesnt appear to be used.

File: model/product_notices.go
Comment (ashishbhate) [inline model/product_notices.go:29]: Here we've documented that the keys are in the format `en-US` but in the code we've used `enUS`

File: app/product_notices.go
Comment (ashishbhate) [inline app/product_notices.go:221]: Doesn't this mean the returned filtered notices will all be in `enUS` instead of the requested locale?

File: app/product_notices.go
Comment (reflog) [reply app/product_notices.go:36]: @ashishbhate  correct. in this phase - locale is ignored

File: app/product_notices.go
Comment (reflog) [reply app/product_notices.go:221]: @ashishbhate correct. in this phase - locale is ignored

**Review by ashishbhate — COMMENTED** [review]
Looks good, some questions about locale inline.

File: model/product_notices.go
Comment (reflog) [reply model/product_notices.go:29]: i'll update the comment

**Review by ashishbhate — APPROVED** [review]
LGTM after comment update.

File: app/product_notices.go
Comment (jespino) [inline app/product_notices.go:36]: Looks like you are receiving an `*App` parameter but you are not using it.

File: app/product_notices.go
Comment (jespino) [inline app/product_notices.go:154]: I don't think this is strictly wrong, but I don't know if this is strictly right either. This is saying that any user that has permissions to manage the system is a system admin, and any user with permissions to manage a team is a team admin. And that is kinda right. But maybe is bit more "right" to say that any user with the "system_admin" role is a system admin, and any user with the "team_admin" role is a team admin. From the notifications point of view, maybe makes more sense one or the other. probably is a question for PMs. The current approach is perfectly fine, but I think the other approach too.

File: app/product_notices.go
Comment (jespino) [inline app/product_notices.go:188]: this inline function looks a bit more complex to understad for me than a regular inline block, somethig like:

```suggestion
		var view *model.ProductNoticeViewState
		for _, v := range views {
			if v.NoticeId == nid {
				view = &v
				break
			}
		}
```

But maybe is matter of taste.

File: app/product_notices.go
Comment (jespino) [inline app/product_notices.go:201]: This feels confusing, why not to use another variable?

File: config/utils.go
Comment (jespino) [inline config/utils.go:185]: Why this wrapper function here?

**Review by jespino — CHANGES_REQUESTED** [review]
Some tiny details to discuss, but in general looks good to me.

File: app/product_notices.go
Comment (reflog) [reply app/product_notices.go:154]: No idea which is the righter one :) Pinging @esethna

File: config/utils.go
Comment (reflog) [reply config/utils.go:185]: @jespino - we didn't have a function for getting config values by dotted path, so I've copied one from mmctl

Comment (jespino) [reply config/utils.go:185]: Yes, still, I don't think the wrapper is providing any value.

File: app/product_notices.go
Comment (reflog) [reply app/product_notices.go:36]: @jespino *App is being used by validateConfigEntry function

Comment (jespino) [reply app/product_notices.go:36]: o yes, I didn't see that, but probably you only need to pass the current config, right? not the whole App

Comment (reflog) [reply app/product_notices.go:36]: yup. i'll fix that for clarity

File: app/product_notices.go
Comment (esethna) [reply app/product_notices.go:154]: For the purposes of notices we want to look at the user role. ie any user with the "system_admin" role is a system admin, and any user with the "team_admin" role is a team admin @reflog

**Comment by esethna** [comment]
Thanks for the PR @reflog, whats our plan for testing so we can get it merged?

**Comment by reflog** [comment]
@jespino - anything else blocking the PR?

**Comment by jespino** [comment]
@reflog nothing blocking, sorry the delay

**Review by jespino — APPROVED** [review]

**Comment by esethna** [comment]
@reflog are we good to merge this?

**Comment by reflog** [comment]
@esethna - not yet. The notices hosting is still unavailable.

File: app/product_notices.go
Comment (jespino) [inline app/product_notices.go:37]: I would prefer to pass the conf here, and pass an `store.Store` parameter too. Passing the app you are passing too much logic, and binding this to the app layer without explicit need.

**Review by jespino — APPROVED** [review]
An improvement suggestion, up to you to cover that, but I think looks better with less coupling with the App layer.

**Review by ogi-m — APPROVED** [review]
**QA approving for merge.**

We verified in-product notices for: 
- All User Roles: Sysadmin, TeamAdmin, Member and All
- Client Type (desktop/web)
- Server Versions (various values)
- Desktop Version (various values)
- Repeatable
- Priority of notices (ordered by json placement)
- SKU

^ and some combinations of the conditions  

Existing Issues:

- Teamadmin notice preference setting (filed new issue for followup)
- Date Bounds - we only tested for today and received a white screen and console error, @reflog is looking into it
- We added few observations/questions to webapp PR from @sudheerDev concerning hand pointer on circle indicators and <previous button

cc @jgilliam17 @esethna

**Review by jgilliam17 — APPROVED** [review]

**Review by esethna — APPROVED** [review]

**Pull request merged**
